edge-modules/metrics-collector/src/IotHubUpload/IotHubUpload.cs

```diff
@@ -59,6 +48,58 @@ public async Task<bool> PublishAsync(IEnumerable<Metric> metrics, CancellationTo
             }
         }
 
+        private byte[] serializeMetrics(IEnumerable<ExportMetric> outputMetrics)
+        {
+            string outputString = JsonConvert.SerializeObject(outputMetrics);
+
+            if (Settings.Current.TransformForIoTCentral)
+            {
+                outputString = Transform(outputMetrics);
+            }
+
+            byte[] metricsData = Encoding.UTF8.GetBytes(outputString);
+
+            if (Settings.Current.CompressForUpload)
+            {
+                metricsData = Compression.CompressToGzip(metricsData);
+            }
+
+            return metricsData;
+        }
+
+        private IEnumerable<Message> BatchAndBuildMessages(List<IEnumerable<ExportMetric>> splitMetrics)
+        {
+
+            List<Message> messageList = new List<Message>();
+            byte[] metricsData = serializeMetrics(splitMetrics.Last());
+            if (metricsData.Length > Constants.MaxMessageSize)
+            {
+                LoggerUtil.Writer.LogInformation($"IoT message is {metricsData.Length} bytes, splitting messages...");
+                List<IEnumerable<ExportMetric>> splitAgainMetrics = new List<IEnumerable<ExportMetric>>();
+                foreach (IEnumerable<ExportMetric> metrics in splitMetrics)
+                {
+                    ExportMetric[] metricsarray = metrics.ToArray();
+                    ExportMetric[] firstArray = metricsarray.Take(metricsarray.Length / 2).ToArray();
+                    splitAgainMetrics.Add(firstArray);
+                    ExportMetric[] secondArray = metricsarray.Skip(metricsarray.Length / 2).ToArray();
+                    splitAgainMetrics.Add(secondArray);
+                }
+
+                return BatchAndBuildMessages(splitAgainMetrics);
+            }
+            else
+            {
+                foreach (IEnumerable<ExportMetric> metrics in splitMetrics)
+                {
+                    metricsData = serializeMetrics(metrics);

```

Comment:
A list of metrics that is close to but not exceeding the size limit will always be serialized twice, including possibly transforming the data for IoT Central and compressing the data. If the list of metrics exceeds the size limit, then the entire list will be serialized, followed by half the list, etc., until all sub-lists are within the size, then each sublist is serialized again. So we're going from O(n) to O(n log n) performance. This will likely be a noticeable (and possibly unacceptable) performance dip for many users.